File: UnityProject/Assets/Simulation/Core/World.cs
using System.Collections.Generic;

namespace HistoryTeller.Simulation
{
    /// <summary>Состояние мира: флаги персонажей + направленные отношения. Чистый C#, без UnityEngine.</summary>
    public sealed class World
    {
        public const string DeadFlag = "dead";

        private readonly Dictionary<string, HashSet<string>> _flags =
            new Dictionary<string, HashSet<string>>();
        private readonly HashSet<(string rel, string from, string to)> _relations =
            new HashSet<(string, string, string)>();

        public bool HasFlag(string charId, string flag) =>
            _flags.TryGetValue(charId, out var s) && s.Contains(flag);

        public void SetFlag(string charId, string flag)
        {
            if (!_flags.TryGetValue(charId, out var s))
                _flags[charId] = s = new HashSet<string>();
            s.Add(flag);
        }

        public void RemoveFlag(string charId, string flag)
        {
            if (_flags.TryGetValue(charId, out var s)) s.Remove(flag);
        }

        public bool HasRelation(string rel, string from, string to) =>
            _relations.Contains((rel, from, to));

        public void AddRelation(string rel, string from, string to) =>
            _relations.Add((rel, from, to));

        public void RemoveRelation(string rel, string from, string to) =>
            _relations.Remove((rel, from, to));

        public bool IsAlive(string charId) => !HasFlag(charId, DeadFlag);

        /// <summary>Глубокая копия (для снапшотов после каждой панели).</summary>
        public World Clone()
        {
            var w = new World();
            foreach (var kv in _flags)
                w._flags[kv.Key] = new HashSet<string>(kv.Value);
            foreach (var r in _relations)
                w._relations.Add(r);
            return w;
        }
    }
}

File: ios/Modules/Simulation/Sources/Simulation/World.swift
import Foundation

/// Направленное отношение между акторами. В C# это был кортеж (rel, from, to) в HashSet;
/// в Swift кортежи не Hashable — заворачиваем в структуру.
struct Relation: Hashable {
    let rel: String
    let from: String
    let to: String
}

/// Состояние мира: флаги персонажей + направленные отношения.
/// Чистый Swift, без UIKit/SwiftUI. Мутируемый — как в оригинальном движке.
public final class World {
    public static let deadFlag = "dead"

    private var flags: [String: Set<String>] = [:]
    private var relations: Set<Relation> = []

    public init() {}

    public func hasFlag(_ charId: String, _ flag: String) -> Bool {
        flags[charId]?.contains(flag) ?? false
    }

    public func setFlag(_ charId: String, _ flag: String) {
        flags[charId, default: []].insert(flag)
    }

    public func removeFlag(_ charId: String, _ flag: String) {
        flags[charId]?.remove(flag)
    }

    public func hasRelation(_ rel: String, _ from: String, _ to: String) -> Bool {
        relations.contains(Relation(rel: rel, from: from, to: to))
    }

    public func addRelation(_ rel: String, _ from: String, _ to: String) {
        relations.insert(Relation(rel: rel, from: from, to: to))
    }

    public func removeRelation(_ rel: String, _ from: String, _ to: String) {
        relations.remove(Relation(rel: rel, from: from, to: to))
    }

    public func isAlive(_ charId: String) -> Bool { !hasFlag(charId, World.deadFlag) }

    /// Глубокая копия (для снапшотов после каждой панели).
    /// Dictionary и Set в Swift — value-типы (copy-on-write), присваивание = глубокое копирование.
    public func clone() -> World {
        let w = World()
        w.flags = flags
        w.relations = relations
        return w
    }
}
